### ISSUE TITLE
amp-base-carousel e2e test is flaky

### ISSUE THREAD

**Issue by lannka** [ampproject/amphtml#24195]
``` 
 1) AMP carousel arrows with custom arrows
       firefox
          Standalone environment 
            
             should go to the next slide:
      AssertionError: expected { Object (bottom, height, ...) } to have property 'x' of 0, but got 800
      + expected - actual
      -800
      +0
```

**Comment by nainar** [comment]
Bumping the Priority down since the test is skipped and the component is experimental

**Comment by rsimha** [comment]
Bumping this issue to P1, since we have newly flaky AMP carousel tests.

Logs from today's `master` builds:

https://travis-ci.org/github/ampproject/amphtml/jobs/673119438#L2311
https://travis-ci.org/github/ampproject/amphtml/jobs/673154489#L2311

```
  1) AMP carousel
       chrome
          Viewer environment 
            
             should have the correct scroll position when resizing:
      AssertionError: expected { Object (bottom, height, ...) } to have property 'x' of 0, but got -900
      + expected - actual
      --900
      +0
      
  
Executed 434 of 484 (Skipped 51) 1 FAILED 
```

**Comment by rsimha** [comment]
@ampproject/wg-components This suite of tests has been failing quite regularly of late and is affecting most CI runs. For example: [this one](https://app.circleci.com/pipelines/github/ampproject/amphtml/199/workflows/3fa375ca-6b63-49fd-9175-85e36bd7eeb0/jobs/1282/parallel-runs/0/steps/0-110).

![image](https://user-images.githubusercontent.com/26553114/105408856-8586f580-5bfd-11eb-808c-4a3ff72309b1.png)

The error is in `beforeEach()` (likely because the carousel arrows aren't appearing on time). Assuming the fix isn't straightforward, okay if I send out a PR to temporarily skip this suite?

**Comment by micajuine-ho** [comment]
@rsimha SGTM. I'll take a look ASAP.

**Comment by caroqliu** [comment]
@rsimha As a point of clarification, the above screenshot looks relevant to 1.0 (bento) amp-base-carousel and may not be related to the other pre-existing flaky tests this issue was filed for. I'll take a look. I see that it is [passing most of the time](http://amp-test-cases.appspot.com/test-cases/stats/fail), is there a better resource that shows the higher rates of failure?

**Comment by rsimha** [comment]
@caroqliu Here is the list of `master` builds on CircleCI: https://app.circleci.com/pipelines/github/ampproject/amphtml?branch=master. If you click through failing builds, most if not all the E2E failures are due to this test suite. (Obviously, feel free to ignore the other failing suites like unit / performance. I'm following up on those separately.)

**Comment by caroqliu** [comment]
Thank you! I'll look briefly and send out a skip PR by EOD if there isn't an obvious fix.

**Comment by caroqliu** [comment]
Some notes after reviewing runs 99-213:
- [`amp-base-carousel test snap property`](https://app.circleci.com/pipelines/github/ampproject/amphtml/213/workflows/cc34f8e7-a425-430f-8563-8ed9dce54142/jobs/1386) is a 0.1 failure with `StaleElementReferenceError: stale element reference: element is not attached to the page document`
- [`AMP carousel arrows when non-looping`](https://app.circleci.com/pipelines/github/ampproject/amphtml/200/workflows/8ac14e2a-b652-4d14-9358-0b9ec022b131/jobs/1232) is a 1.0 failure with `TimeoutError: Waiting for element to be located .arrow-prev-34d526e`
- [`AMP carousel should snap when scrolling`](https://app.circleci.com/pipelines/github/ampproject/amphtml/190/workflows/ecc4511b-dfa2-45d9-9ea4-adab88fa2651/jobs/1178) is a 1.0 failure with `TimeoutError: Waiting for element to be located .hide-scrollbar-34d526e`
- [`AMP carousel should snap on center point`](https://app.circleci.com/pipelines/github/ampproject/amphtml/128/workflows/4e02b739-e69d-49d9-af61-4575c4b12345/jobs/689) is a 1.0 failure with `TimeoutError: Waiting for element to be located .hide-scrollbar-34d526e`

@micajuine-ho want to investigate the stale element reference in 0.1?

For the others I wonder if increasing the timeouts for those test files will help. I'm not able to reproduce any of them locally unfortunately and wonder if it has to do with the processing speed of the test runner. :/ Are Circle CI tests still non-blocking? Would it be possible to increase the timeout for the next day to see if it helps, and skip otherwise?

Also for more obvious version differentiation I'll send a PR to update the test names to include their versions.

**Comment by rsimha** [comment]
Wow, that was comprehensive. Thanks for the careful analysis, @caroqliu!

> For the others I wonder if increasing the timeouts for those test files will help. I'm not able to reproduce any of them locally unfortunately and wonder if it has to do with the processing speed of the test runner. :/ Are Circle CI tests still non-blocking? Would it be possible to increase the timeout for the next day to see if it helps, and skip otherwise?

We're currently using the "XLarge" VM provided by CircleCI, so tests should run fairly fast.

https://github.com/ampproject/amphtml/blob/ca87c8e7f762964c5bd21aeab438d1b42d051aa2/.circleci/config.yml#L21-L25

CircleCI tests are indeed non-blocking for now (as is the `ampproject/tests/e2e (nomodule)` test status). The plan is to make it blocking in the near future and switch off of Travis. Not opposed to increasing the timeouts, since a few extra seconds can help us avoid multiple minutes of re-running time.

**Comment by rsimha** [comment]
> * [`amp-base-carousel test snap property`](https://app.circleci.com/pipelines/github/ampproject/amphtml/213/workflows/cc34f8e7-a425-430f-8563-8ed9dce54142/jobs/1386) is a 0.1 failure with `StaleElementReferenceError: stale element reference: element is not attached to the page document`

I have more data points after observing weekend builds. The remaining flake is from the 0.1 tests, and likely has to do with initial state. This might need a fix for the `beforeEach()`, since I'm not sure skipping individual tests is sufficient. For example, [this build](https://app.circleci.com/pipelines/github/ampproject/amphtml/385/workflows/56ac39b5-04c4-4683-8b72-e0d893d25a2b/jobs/3178):
```js
  1) amp-base-carousel:0.1 - arrows when non-looping
       chrome
          Viewer environment 
           "before each" hook for "should have the arrows in the correct initial state":
     Error: Server terminated early with status 1
      at /home/circleci/project/build-system/tasks/e2e/node_modules/selenium-webdriver/remote/index.js:248:24
      at processTicksAndRejections (internal/process/task_queues.js:93:5)

  2) amp-base-carousel:0.1 - arrows when non-looping
       chrome
          Viewer environment 
           "after each" hook for "should have the arrows in the correct initial state":
     Error: Server terminated early with status 1
      at /home/circleci/project/build-system/tasks/e2e/node_modules/selenium-webdriver/remote/index.js:248:24
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

**Comment by caroqliu** [comment]
As @rcebulko kindly pointed out, these [appear to be flaking again](http://amp-test-cases.appspot.com/test-cases/stats/fail). Will take a look.

**Comment by rcebulko** [comment]
> As @rcebulko kindly pointed out, these [appear to be flaking again](http://amp-test-cases.appspot.com/test-cases/stats/fail). Will take a look.

Posting screenshot for posterity/reference since that link may become outdated
![image](https://user-images.githubusercontent.com/6694512/111181125-07532980-8584-11eb-9fed-c2dee138ed8f.png)

**Comment by rsimha** [comment]
Tagging @rileyajones as an FYI because he has been investigating other non-specific E2E failures via #32442.

**Comment by raja-anbazhagan** [comment]
Umm, My PR #33383 fails for `should have the correct scroll position when resizing` test case.

```txt
 1) [v1.0] amp-base-carousel - non-looping
       chrome
          Standalone environment 
            
             should have the correct scroll position when resizing:
     TimeoutError: Waiting for element to be located .slide-element-34d526e[data-slide="0"]
Wait timed out after 5130ms
      at /home/circleci/project/build-system/tasks/e2e/node_modules/selenium-webdriver/lib/webdriver.js:892:17
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

Is this anyway related/similar?

**Comment by caroqliu** [comment]
Note: The following is only applicable to amp-base-carousel 1.0 and other (Preact-based) Bento components:

After the recently flaking amp-lightbox:1.0 tests (related #33414), @dvoytenko mentioned that it's possible `useEffect` cleanup functions might be running a bit later, causing the component state and the testing requirements to go out of sync. Most of non-Bento elements are essentially never cleaned up. But Bento always cleans up after itself, meaning there could be side effects we are not precisely coordinating. 

In particular the flakes usually come in the form of one of two errors:

(1) `TimeoutError: Waiting for element to be located `
(2) `StaleElementReferenceError: stale element reference: element is not attached to the page document`

This got me thinking -- perhaps we have assumed that our e2e testing infrastructure is fundamentally compatible with React when it actually isn't, or at least requires some more nuanced application of our existing tools.

Since Bento's Preact components introduced the need to include `enzyme` for unit tests, it's possible that we could benefit from doing something similar for e2e tools as well. That led me to this open issue in the enzyme repo (https://github.com/enzymejs/enzyme/issues/237) which ends in this resource here: https://github.com/DianaSuvorova/e2e

In particular, the problem described in the above resource appears on the surface to be very close to ours: 
> In selenium tests DOM elements normally get looked up on the page with css selectors or by x-path. And this is the tricky part - it is error-prone (the DOM structure changes a often) and with the javascript inline styles there is no real need to add classes or ids except for purpose of testing.
>
> Also with React Apps DOM elements is no longer the right level of abstraction. Ideally one should be able to get an access to a React element on the page and within it grab underlying DOM elements to trigger events and assert values.

It's possible there's something we could learn from here given our recent pain points with flakiness -- could use some guidance from infra if this sounds like a worthy direction to explore.

**Comment by rcebulko** [comment]
Another flake: https://app.circleci.com/pipelines/workflows/eebdd85b-86fc-448b-a3b9-3f33af6a0e6f

![image](https://user-images.githubusercontent.com/6694512/112177692-86bdaa00-8bcf-11eb-9752-fcc045fb7be1.png)

**Comment by Enriqe** [comment]
More flaky tests are blocking main from being green:

https://app.circleci.com/pipelines/github/ampproject/amphtml/9854/workflows/e0bd0a95-2391-49cf-b7ce-2ad4e236b4af/jobs/152844

**Comment by caroqliu** [comment]
> > More flaky tests are blocking main from being green:
> > https://app.circleci.com/pipelines/github/ampproject/amphtml/9854/workflows/e0bd0a95-2391-49cf-b7ce-2ad4e236b4af/jobs/152844
> 
> It doesn't look like these tests are flaky... it seems like they are failing pretty consistently. And I am not seeing an obvious culprit in the recent commit history. It will require further investigation. Skip PR on the way!

Skipped in #34541. When I went to block, I wasn't able to reproduce the failures locally (running `amp e2e --files=`). We will have to be able to get them to fail on test runs to investigate, fix, and unskip these tests without re-blocking master on accident. @rileyajones is it possible to run tests on Circle CI for investigation purposes? I suppose we could do it via pushing commits to a PR, but that would be a slow roundtrip if we are just testing out changes. Any advice would be much appreciated :)
